### ISSUE TITLE
pretty-format-toml removes trailing comments in list

### ISSUE THREAD

**Issue by l0b0** [macisamuele/language-formatters-pre-commit-hooks#202]
When reformatting a list setting like

```toml
load-plugins = [
  "pylint.extensions.bad_builtin",
  # "pylint.extensions.docparams",
  "pylint.extensions.set_membership"
  # "pylint.extensions.typing"
```

, the first comment is retained, but the second one is removed. Both comments should be retained. Not sure whether this is an issue with TomlSort or this repo.

**Comment by maresb** [comment]
This looks like a duplicate of https://github.com/macisamuele/language-formatters-pre-commit-hooks/issues/161

**Comment by maresb** [comment]
I feel like this TOML pre-commit hook needs a big red warning that it's unsafe since it will unceremoniously delete comments.

**Comment by Avasam** [comment]
I find it'll do so even outside lists. And I have no workaround, making this hook pretty unusable

**Comment by corneliusroemer** [comment]
Yes, TOML needs to be fixed or marked unsafe as it removes comments - don't use!

<img width="1727" alt="image" src="https://github.com/user-attachments/assets/4a0f8b3a-360c-4766-9839-c5a4b4fa905c">

**Comment by maresb** [comment]
@corneliusroemer I don't see a corresponding PR for the commit linked above. Is one on the way, or did I miss it? Thanks for following through with that!

**Comment by corneliusroemer** [comment]
@maresb That is for reminding me! I must have forgotten to make a PR. Done now! #242

**Comment by maresb** [comment]
By the way, if people need a safe TOML pre-commit hook, this just became available a few days ago and it seems to work very well. (The previous version used a cumbersome Docker contaier.):

https://github.com/ComPWA/mirrors-taplo

**Comment by macisamuele** [comment]
Unfortunately, the hook is not unsafe.
For a formatter unsafe is when the generated output is semantically different than the input.

In the issue a few points were mentioned:
* removal of trailing comma: the semantic of the content is not altered, hence the difference is about rendering
* removal of comment not bound to any item: the semantic of the content is not altered

I could understand the argument that there might be better base library to use, and I would be welcoming PRs in this direction.

**Comment by l0b0** [comment]
>     * removal of comment not bound to any item: the semantic of the content is not altered

Semantics isn't only about computers, though. Removing a comment definitely changes the semantics for any human readers.

**Comment by Avasam** [comment]
If it accidentally dropped comments whilst applying a fix it'd be one thing, since you can add them back and the formatting should be stable after that. But it currently removes comments without applying any real changes.

Semantics of "unsafe" aside, the end result is I can't use this hook because it can't keep some comments.

**Comment by maresb** [comment]
@macisamuele, the issue of `toml-sort` deleting comments definitely isn't your fault. It's an upstream issue https://github.com/pappasam/toml-sort/issues/59 so I hope you are not taking this personally.

On the other hand, many people (including me) store critical contextual information in comments. I am unfamiliar with any popular formatter (other than `toml-sort`) that deletes comments in its default configuration, although I'm happy to be corrected on this point. Therefore, I hope that you would agree that this hook is "unexpectedly destructive" in the most literal sense possible. It is unexpected in the sense that it is extremely uncommon for a formatter to delete comments and has caught me and all the above commenters offguard. It is destructive in that I often put significant work and information into my comments, they often contain information and TODOs, and the disappearance of this information would eventually likely result in product downtime.

As such, I have opened #249 as a follow-up to #242 as neutral statement of facts.

**Comment by maresb** [comment]
Thanks @macisamuele for merging #249. There's now a warning in the README:


## List of pretty-format hooks
* `pretty-format-golang`
* `pretty-format-ini`
* `pretty-format-java`
* `pretty-format-kotlin`
* `pretty-format-rust`
* `pretty-format-toml` ⚠️ removes comments under certain conditions
* `pretty-format-yaml`
